rand_distr/src/zipf.rs

```diff
@@ -0,0 +1,328 @@
+            if v * x * (t - F::one()) / (self.b - F::one()) <= t / self.b {
+                return x;
+            }
+        }
+    }
+}
+
+/// Samples integers according to the Zipf distribution.
+///
+/// The samples follow Zipf's law: The frequency of each sample from a finite
+/// set of size `n` is inversely proportional to a power of its frequency rank
+/// (with exponent `s`).
+///
+/// For large `n`, this converges to the [`Zeta`] distribution.
+///
+/// For `s = 0`, this becomes a uniform distribution.
+///
+/// # Example
+/// ```
+/// use rand::prelude::*;
+/// use rand_distr::Zipf;
+///
+/// let val: f64 = thread_rng().sample(Zipf::new(10, 1.5).unwrap());
+/// println!("{}", val);
+/// ```
+#[derive(Clone, Copy, Debug)]
+pub struct Zipf<F>
+where F: Float, Standard: Distribution<F> {
+    n: F,
+    s: F,
+    t: F,
+}
+
+/// Error type returned from `Zipf::new`.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum ZipfError {
+    /// `s < 0` or `nan`.
+    STooSmall,
+    /// `n < 1`.
+    NTooSmall,
+}
+
+impl fmt::Display for ZipfError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(match self {
+            ZipfError::STooSmall => "s < 0 or is NaN in Zipf distribution",
+            ZipfError::NTooSmall => "n < 1 in Zipf distribution",
+        })
+    }
+}
+
+#[cfg(feature = "std")]
+#[cfg_attr(doc_cfg, doc(cfg(feature = "std")))]
+impl std::error::Error for ZipfError {}
+
+impl<F> Zipf<F>
+where F: Float, Standard: Distribution<F> {
+    /// Construct a new `Zipf` distribution for a set with `n` elements and a
+    /// frequency rank exponent `s`.
+    ///
+    /// For large `n`, rounding may occur to fit the number into the float type.
+    #[inline]
+    pub fn new(n: u64, s: F) -> Result<Zipf<F>, ZipfError> {
+        if !(s >= F::zero()) {
+            return Err(ZipfError::STooSmall);
+        }
+        if n < 1 {
+            return Err(ZipfError::NTooSmall);
+        }
+        let n = F::from(n).unwrap();  // This does not fail.
+        let t = if s != F::one() {
+            (n.powf(F::one() - s) - s) / (F::one() - s)
+        } else {
+            F::one() + n.ln()
+        };
+        Ok(Zipf {
+            n, s, t
+        })
+    }
+
+    /// Inverse cumulative density function
+    #[inline]
+    fn inv_cdf(&self, p: F) -> F {
+        let one = F::one();
+        let pt = p * self.t;
+        if pt <= one {
+            pt
+        } else if self.s != F::one() {
+            (pt * (one - self.s) + self.s).powf(one / (one - self.s))
+        } else {
+            pt.exp()
+        }
+    }
+}
+
+impl<F> Distribution<F> for Zipf<F>
+where F: Float, Standard: Distribution<F>
+{
+    #[inline]
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> F {
+        let one = F::one();
+        loop {
+            let inv_b = self.inv_cdf(rng.sample(Standard));
+            let x = (inv_b + one).floor();
+            let mut ratio = x.powf(-self.s) * self.t;
```

Comment:
I do not think there is the need to multiply by `self.t` here.
If you look at the [java code](https://jasoncrease.medium.com/rejection-sampling-the-zipf-distribution-6b359792cffa), `t` actually does not appear in the ratio. Or more specifically, it is simplified (the code can be made more efficient without it).

Here is the reasoning.
- Assume we want to sample from a continuous random variable `Y` whose distribution is proportional to `f(y) = ceil(y)^{-s}`, for `y > 0`.
- We have another density `b` such that `f(y) <= t * b(y)` for all `y > 0`.
- The rejection sampling says that, once we have a candidate `z` sampled from density `b`, the ratio of acceptance should be `RATIO = f(z) / (t * b(z))`.
- Notice that the definition of `b` makes it so that
```
t * b(z) = if z in (0, 1] { 1 }
else { z^{-s} }
```
- Therefore, 
```
RATIO = f(z) / (t * b(z)) = 
if z in (0, 1] { ceil(z)^{-s} } // This is incorrect in the code, no t needed
else { ceil(z)^{-s} * z^s } // This is incorrect in the code
```
- If accepted, `Z` will follow `f`, effectively sampling from `Y`. In our case, we do not care about the continuous random variable `Y`, but about `ceil(Y)`.

In the code, `ceil(z)` is `x`, while `z` is `inv_b`.

Reply:
You're right, the `t` in the reference implementation cancels (for `t != 0`). I'll fix that.